packages/tools/src/tools/annotation/SplineROITool.ts

```diff
@@ -0,0 +1,1107 @@
+  ) {
+    const enabledElement = getEnabledElement(element);
+    const { points: controlPoints } = annotation.data.handles;
+
+    // There is no curve with only 2 points
+    if (controlPoints.length === 3) {
+      removeAnnotation(annotation.annotationUID);
+    } else {
+      controlPoints.splice(controlPointIndex, 1);
+    }
+
+    const { renderingEngine } = enabledElement;
+    const viewportIdsToRender = getViewportIdsWithToolToRender(
+      element,
+      this.getToolName()
+    );
+
+    annotation.invalidated = true;
+
+    triggerAnnotationRenderForViewportIds(renderingEngine, viewportIdsToRender);
+  }
+
+  deleteControlPointCallback = (
+    evt: EventTypes.InteractionEventType,
+    annotation: SplineROIAnnotation
+  ) => {
+    const splineType = annotation.data.spline.type;
+    const splineConfig = this._getSplineConfig(splineType);
+    const maxDist = splineConfig.controlPointDeletionDistance;
+
+    if (splineConfig.controlPointDeletionEnabled === false) {
+      return;
+    }
+
+    const eventDetail = evt.detail;
+    const { element, currentPoints } = eventDetail;
+    const { canvas: canvasPos } = currentPoints;
+    const spline = this._getAnnotationSpline(element, annotation);
+    const closestControlPoint = spline.getClosestControlPointWithinRange(
+      canvasPos,
+      maxDist
+    );
+
+    if (!closestControlPoint) {
+      return;
+    }
+
+    this._deleteControlPointByIndex(
+      element,
+      annotation,
+      closestControlPoint.index
+    );
+  };
+
+  /**
+   * Get a spline config merged with the default settings.
+   * @param type - Spline type (CARDINAL, CATMULLROM, LINEAR or BSPLINE)
+   * @returns Spline configuration
+   */
+  private _getSplineConfig(type: string) {
+    const { configuration: config } = this;
+    const splineConfigs = config.spline.configuration;
+
+    return Object.assign({ type }, DEFAULT_SPLINE_CONFIG, splineConfigs[type]);
+  }
+
+  private _updateSplineScale(spline: ISpline, annotation: SplineROIAnnotation) {
+    const splineType = annotation.data.spline.type;
+    const splineConfig = this._getSplineConfig(splineType);
+
+    if (
+      !(spline instanceof CardinalSpline) ||
+      spline.fixedScale ||
+      splineConfig.scale === undefined ||
+      spline.scale === splineConfig.scale
+    ) {
+      return;
+    }
+
+    spline.scale = splineConfig.scale;
+    annotation.invalidated = true;
+  }
+
+  private _createOrUpdateAnnotationSpline = (
+    element: HTMLDivElement,
+    annotation: SplineROIAnnotation
+  ): ISpline => {
+    const enabledElement = getEnabledElement(element);
+    const { viewport } = enabledElement;
+    const { worldToCanvas } = viewport;
+    const splineType = annotation.data.spline.type;
+    const splineConfig = this._getSplineConfig(splineType);
+    let spline = this.splineInstancesCache.get(annotation);
+
+    if (!spline) {
+      spline = new splineConfig.Class();
+      this.splineInstancesCache.set(annotation, spline);
```

Comment:
I believe it would be better to have the spline instance caches inside the annotation itself for a couple of reasons. Firstly, it would be more manageable for memory. Secondly, it would align better with our architecture.